pkg/filters/grpcproxy/proxy.go

```diff
@@ -0,0 +1,309 @@
+			return fmt.Errorf("pool %d: %v", i, err)
+		}
+	}
+
+	if numMainPool != 1 {
+		return fmt.Errorf("one and only one mainPool is required")
+	}
+
+	var (
+		borrowTimeout  time.Duration
+		err            error
+		connectTimeout time.Duration
+	)
+
+	if s.BorrowTimeout != "" {
+		if borrowTimeout, err = time.ParseDuration(s.BorrowTimeout); err != nil || borrowTimeout == 0 {
+			return fmt.Errorf("grpc client borrow timeout %s invalid", s.BorrowTimeout)
+		}
+	}
+	if s.ConnectTimeout != "" {
+		if connectTimeout, err = time.ParseDuration(s.ConnectTimeout); err != nil || connectTimeout == 0 {
+			return fmt.Errorf("grpc client wait connection ready timeout %s invalid", s.ConnectTimeout)
+		}
+
+	}
+	if borrowTimeout <= connectTimeout {
+		return fmt.Errorf("grpc client borrow connection timeout %v should greater than connect timeout %v", borrowTimeout, connectTimeout)
+	}
+
+	return nil
+}
+
+// Name returns the name of the Proxy filter instance.
+func (p *Proxy) Name() string {
+	return p.spec.Name()
+}
+
+// Kind returns the kind of Proxy.
+func (p *Proxy) Kind() *filters.Kind {
+	return kind
+}
+
+// Spec returns the spec used by the Proxy
+func (p *Proxy) Spec() filters.Spec {
+	return p.spec
+}
+
+// Init initializes Proxy.
+func (p *Proxy) Init() {
+	p.reload()
+}
+
+// Inherit inherits previous generation of Proxy.
+func (p *Proxy) Inherit(previousGeneration filters.Filter) {
+	p.reload()
+}
+
+func (p *Proxy) reload() {
+	for _, spec := range p.spec.Pools {
+		name := ""
+		if spec.Filter == nil {
+			name = fmt.Sprintf("proxy#%s#main", p.Name())
+		} else {
+			id := len(p.candidatePools)
+			name = fmt.Sprintf("proxy#%s#candidate#%d", p.Name(), id)
+		}
+
+		pool := NewServerPool(p, spec, name)
+
+		if spec.Filter == nil {
+			p.mainPool = pool
+		} else {
+			p.candidatePools = append(p.candidatePools, pool)
+		}
+	}
+
+	p.closeEvent = make(chan struct{}, 1)
+
+	if !p.spec.UseConnectionPool {
+		p.conns = make(map[string]*Conn)
+		go p.monitor()
+	} else {
+		// already valid in Spec.Validate()
+		borrowTimeout, _ := time.ParseDuration(p.spec.BorrowTimeout)
+		connectTimeout, _ := time.ParseDuration(p.spec.ConnectTimeout)
+		poolSpec := &grpcpool.Spec{
+			InitConnNum:     p.spec.InitConnNum,
+			BorrowTimeout:   borrowTimeout,
+			ConnectTimeout:  connectTimeout,
+			MaxConnsPerHost: p.spec.MaxConnsPerHost,
+			DialOptions:     []grpc.DialOption{grpc.WithBlock(), grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithCodec(grpcpool.GetCodecInstance())},
+		}
+
+		p.pool = grpcpool.NewPool(poolSpec)
+	}
+
+}
+
+// Status returns Proxy status.
+func (p *Proxy) Status() interface{} {
+	return nil
+}
+
+// Close closes Proxy.
+func (p *Proxy) Close() {
+	p.mainPool.close()
+
+	for _, v := range p.candidatePools {
+		v.close()
+	}
+
+	// help gc
+	if !p.spec.UseConnectionPool {
+		p.conns = nil
+	} else {
+		p.pool.Close()
+		p.pool = nil
+	}
+	p.closeEvent <- struct{}{}
+	close(p.closeEvent)
```

Comment:
for this purpose, we can simply use an un-buffered channel and close it directly.